app/src/main/cpp/skyline/services/hosbinder/GraphicBufferProducer.cpp

```diff
@@ -1,214 +1,702 @@
+                if (bufferSlot == queue.end() || it->frameNumber < bufferSlot->frameNumber)
+                    bufferSlot = it;
+            }
+        }
+
+        if (bufferSlot == queue.end())
+            return AndroidStatus::NoMemory;
 
-        queue.at(slot)->status = BufferStatus::Free;
+        bufferSlot->state = BufferState::Free;
+        bufferSlot->frameNumber = std::numeric_limits<u32>::max();
+        graphicBuffer = *std::exchange(bufferSlot->graphicBuffer, nullptr);
+        fence = AndroidFence{};
 
-        state.logger->Debug("Slot: {}", slot);
+        bufferEvent->Signal();
+
+        state.logger->Debug("#{}", std::distance(queue.begin(), bufferSlot));
+        return AndroidStatus::Ok;
     }
 
-    void GraphicBufferProducer::Connect(Parcel &out) {
-        struct {
-            u32 width{constant::HandheldResolutionW}; //!< The width of the display
-            u32 height{constant::HandheldResolutionH}; //!< The height of the display
-            u32 transformHint{}; //!< A hint of the transformation of the display
-            u32 pendingBuffers{}; //!< The number of pending buffers
-            u32 status{}; //!< The status of the buffer queue
-        } data{};
-        out.Push(data);
-        state.logger->Debug("Connect");
+    AndroidStatus GraphicBufferProducer::AttachBuffer(i32 &slot, const GraphicBuffer &graphicBuffer) {
+        std::scoped_lock lock(mutex);
+        auto bufferSlot{queue.end()};
+        for (auto it{queue.begin()}; it != queue.end(); it++) {
+            if (it->state == BufferState::Free) {
+                if (bufferSlot == queue.end() || it->frameNumber < bufferSlot->frameNumber)
+                    bufferSlot = it;
+            }
+        }
+
+        if (bufferSlot == queue.end()) {
+            state.logger->Warn("Could not find any free slots to attach the graphic buffer to");
+            return AndroidStatus::NoMemory;
+        }
+
+        if (graphicBuffer.magic != GraphicBuffer::Magic)
+            throw exception("Unexpected GraphicBuffer magic: 0x{} (Expected: 0x{})", graphicBuffer.magic, GraphicBuffer::Magic);
+        else if (graphicBuffer.intCount != sizeof(NvGraphicHandle) / sizeof(u32))
+            throw exception("Unexpected GraphicBuffer native_handle integer count: 0x{} (Expected: 0x{})", graphicBuffer.intCount, sizeof(NvGraphicHandle));
+
+        auto &handle{graphicBuffer.graphicHandle};
+        if (handle.magic != NvGraphicHandle::Magic)
+            throw exception("Unexpected NvGraphicHandle magic: {}", handle.surfaceCount);
+        else if (handle.surfaceCount < 1)
+            throw exception("At least one surface is required in a buffer: {}", handle.surfaceCount);
+        else if (handle.surfaceCount > 1)
+            throw exception("Multi-planar surfaces are not supported: {}", handle.surfaceCount);
+
+        auto &surface{graphicBuffer.graphicHandle.surfaces.at(0)};
+        if (surface.scanFormat != NvDisplayScanFormat::Progressive)
```

Comment:
540i interlaced CRT skyline :(

Reply:
I'll let you get Khronos Group to support it in Vulkan first then I can add support for interlaced output to Skyline